File: elixir/triangle/triangle.exs
defmodule Triangle do
  @type kind :: :equilateral | :isosceles | :scalene

  @doc """
  Return the kind of triangle of a triangle with 'a', 'b' and 'c' as lengths.
  """
  @spec kind(number, number, number) :: { :ok, kind } | { :error, String.t }
  def kind(a, b, c) when a > 0 and b > 0 and c > 0, do: inequality_kind(a, b, c)
  def kind(_, _, _), do: { :error, "all side lengths must be positive" }

  defp inequality_kind(a, b, c) when a + b > c and a + c > b and b + c > a,
    do: valid_kind(a, b, c)
  defp inequality_kind(_, _, _),
    do: { :error, "side lengths violate triangle inequality" }

  defp valid_kind(a, a, a), do: { :ok, :equilateral }
  defp valid_kind(a, b, c) when a == b or b == c or a == c, do: { :ok, :isosceles }
  defp valid_kind(a, b, c), do: { :ok, :scalene }
end

File: javascript/triangle/triangle.js
'use strict';

class Triangle {
  constructor(a, b, c) {
    this.a = a;
    this.b = b;
    this.c = c;
  }

  kind() {
    const a = this.a;
    const b = this.b;
    const c = this.c;

    assertSides(a, b, c);
    assertInequality(a, b, c);

    return sidesToKinds(a, b, c);
  }
}

const UNIQUE_SIDES_KINDS = {
  1: 'equilateral',
  2: 'isosceles',
  3: 'scalene'
};

function sidesToKinds(a, b, c) {
  return UNIQUE_SIDES_KINDS[uniqueSideCount(a, b, c)];
}

function uniqueSideCount(a, b, c) {
  return new Set([a, b, c]).size;
}

function assertSides(a, b, c) {
  if (a <= 0 || b <= 0 || c <= 0) {
    throw 'Illegal triangle';
  }
}

function assertInequality(a, b, c) {
  const sorted = [a, b, c].sort();
  if (sorted[0] + sorted[1] <= sorted[2]) {
    throw 'Violated triangle inequality';
  }
}

module.exports = Triangle;
